src/Common/src/TypeSystem/Common/CastingHelper.cs

```diff
@@ -0,0 +1,380 @@
+                TypeDesc targetArg = instantiationTarget[i];
+                
+                if (arg != targetArg)
+                {
+                    GenericParameterDesc openArgType = (GenericParameterDesc)instantiationOpen[i];
+
+                    if (openArgType.IsCovariant)
+                    {
+                        if (!arg.IsBoxedAndCanCastTo(targetArg, protect))
+                            return false;
+                    }
+                    else if (openArgType.IsContravariant)
+                    {
+                        if (!targetArg.IsBoxedAndCanCastTo(arg, protect))
+                            return false;
+                    }
+                    else
+                    {
+                        // non-variant
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static bool CanCastToClass(this TypeDesc thisType, TypeDesc otherType, StackOverflowProtect protect)
+        {
+            TypeDesc curType = thisType;
+
+            // If the target type has variant type parameters, we take a slower path
+            if (curType.HasVariance)
+            {
+                // First chase inheritance hierarchy until we hit a class that only differs in its instantiation
+                do
+                {
+                    if (curType == otherType)
+                    {
+                        return true;
+                    }
+
+                    if (curType.CanCastByVarianceToInterfaceOrDelegate(otherType, protect))
+                    {
+                        return true;
+                    }
+
+                    curType = curType.BaseType;
+                }
+                while (curType != null);
+            }
+            else
+            {
+                // If there are no variant type parameters, just chase the hierarchy
+
+                // Allow curType to be nullable, which means this method
+                // will additionally return true if curType is Nullable<T> && (
+                //    currType == otherType
+                // OR otherType is System.ValueType or System.Object)
+
+                // Always strip Nullable from the otherType, if present
+                if (otherType.IsNullable && !curType.IsNullable)
+                {
+                    return thisType.CanCastTo(otherType.Instantiation[0]);
+                }
+
+                do
+                {
+                    if (curType == otherType)
+                        return true;
+
+                    curType = curType.BaseType;
+                } while (curType != null);
+            }
+
+            return false;
+        }
+
+        private static bool IsBoxedAndCanCastTo(this TypeDesc thisType, TypeDesc otherType, StackOverflowProtect protect)
+        {
+            TypeDesc fromUnderlyingType = thisType.UnderlyingType;
+
+            if (IsObjRef(fromUnderlyingType.Category))
```

Comment:
Nit - this could be `... .IsObjRef()` since the rest of the file uses the extension methods.